### PR TITLE
WIP Initial prototype for a safer s3 output committer

### PR DESCRIPTION
**Pull request by mariusvniekerk**
Sketch implementation for some of #2923

### CODE REVIEW

File: dask/bytes/s3.py [hunk 1]
```diff
@@ -62,5 +86,10 @@ def _get_pyarrow_filesystem(self):
         import pyarrow as pa
         return pa.filesystem.S3FSWrapper(self)
 
+    def commit(self, metadata):
+        self._call_s3(
+            self.s3.complete_multipart_upload,
+            **metadata)
```
Comment (mrocklin) [inline dask/bytes/s3.py]: Interesting, I didn't know that this existed.  Does this often exist in other systems as well?  HDFS, GCS, etc..

Comment (mariusvniekerk) [reply dask/bytes/s3.py]: The model for HDFS is typically to write things to a subdirectory and move them to their final location.  I would hope that most of the other blob stores support something similar.

**Comment by mrocklin** [comment]
Pinging @martindurant again in case he has time (which, given his current activities, he may not).

**Comment by martindurant** [comment]
Do I understand the flow correctly:

- when a file is closed, it doesn't call complete-multipart, but writes these into a list for later
- an entire write operation calls _commit_writes and this triggers the creation of the final files almost in one atomic step.

I note that this is only implemented for writing with fastparquet - presumably every output function would need it, so it is an essential property of the file-system backends. Even local-files could have a commit step. How, in the general case, can we ensure it gets called after every set of file outputs?

For the specific case of s3, does this code cope with files that would have been written in a single upload previously, i.e., without multipart?

**Comment by mariusvniekerk** [comment]
So for the specific S3 case.  Yes it delays the commit until a later step.  Files uploaded without multipart don't have any commit stage so there is no near atomic mode.

This is currently only implemented for fastparquet mostly as a way to get a discussion started on output committers. If we decide to go forward with this I'll change this more globally. 

I think in the general case command that write io should optionally include the commit phase at the end or emit the neccesary commit Metadata to potentially commit multiple seperate write calls

**Comment by jsignell** [comment]
I am closing this since it is pretty stale at this point. Please reopen if you are planning on continuing the work.

**Comment by martindurant** [comment]
@alimanfoo , this old conversation might interest you.

**Pull request closed without merging**